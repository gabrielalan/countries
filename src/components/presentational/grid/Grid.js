import React, { useState } from 'react';
import PropTypes from 'prop-types';
import ContainerCell from './ContainerCell';
import ResizableHeader from './headers/ResizableHeader';
import Icon from '../../presentational/icon/Icon';
import { classNames } from '../../../functions/elementHelpers';

/**
 * The Row isn't necessarily complex, so it is included
 * here together with the Grid since for now is really simple.
 * The Row is basically a "bypasser" for the column specific logic.
 */
const createRowRenderer = (columns, indexColumn, Cell) => {
  const columnKeys = Object.keys(columns);

  return ({ data, dimensions }) => <div className="grid__row" key={data[indexColumn]}>
    {columnKeys.map(key => <Cell key={key} width={dimensions[key]} columnSpec={columns[key]} data={data[key]} />)}
  </div>;
};

const createInitialDimensions = columnSpecs => Object.keys(columnSpecs).reduce(
  (acc, name) => Object.assign(acc, { [name]: columnSpecs[name].width }),
  {}
);

/**
 * This Grid is create to render the data in a way
 * that can connect different types of "visualization".
 * The type of cells for instance, can be changed quite easily.
 */
export default function Grid({ indexColumn, columns, data, onColumnHide, onColumnOrder }) {
  const [dimensions, setDimensions] = useState(createInitialDimensions(columns));

  const rowRenderer = createRowRenderer(columns, indexColumn, ContainerCell);

  const setColumnWidth = key => width => setDimensions({
    ...dimensions,
    [key]: width
  });

  const headers = Object.keys(columns).map(key => {
    const column = columns[key];
    const hideButton = column.isHidable ? <button type="button" className="grid__row__column__hide-button" onClick={() => onColumnHide(key)}>âœ–</button> : null;
    const orderIndicator = column.order ? <Icon className="grid__row__column__order" id={column.order === 'ASC' ? 'chevron-bottom' : 'chevron-top'} /> : null;

    return <ResizableHeader onResize={setColumnWidth(key)} initialWidth={column.width} key={key}>
      <button type="button"
        className={classNames({ 'grid__row__column__header-label': true, 'grid__row__column__header-label--is-orderable': column.isOrdarable })}
        onClick={() => column.isOrdarable && onColumnOrder(key, column.order === 'ASC' ? 'DESC' : 'ASC')}>
        {column.label}
        {orderIndicator}
      </button>
      {hideButton}
    </ResizableHeader>;
  });

  const rows = data.map(data => rowRenderer({ data, dimensions }));

  return <section className="grid">
    <div className="grid__row grid__row--header">{headers}</div>
    {rows}
  </section>
}

Grid.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  indexColumn: PropTypes.string.isRequired,
  columns: PropTypes.objectOf(PropTypes.shape({
    Type: PropTypes.any.isRequired,
    width: PropTypes.number.isRequired,
    label: PropTypes.string.isRequired,
    hide: PropTypes.bool,
  })).isRequired
};