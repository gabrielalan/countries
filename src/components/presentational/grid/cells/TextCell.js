/**
 * This is the most basic type of cell used on grids
 * It turns everything into strings so no error will be thrown
 * when the children is some kind of object not accepted by React
 */
export default function TextCell({ children }) {
  return children && children.toString();
}