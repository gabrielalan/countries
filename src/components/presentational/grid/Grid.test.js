import React from 'react';
import renderer from 'react-test-renderer';
import Grid from './Grid';
import TextCell from './cells/TextCell';

describe('Grid', () => {
  it('should render correct number of headers, rows and columns', () => {
    const data = [{
      col1: 'Test 1',
      col2: 1234,
      col3: [1, 2, 3],
      colNotRendered: 'SHOULD NOT BE IN THE SNAPSHOT'
    }, {
      col1: 'Test 2',
      col2: 3421,
      col3: [5, 5, 5],
      colNotRendered: 'SHOULD NOT BE IN THE SNAPSHOT'
    }];

    const columns = {
      col1: {
        Type: TextCell,
        width: 100,
        label: 'Column 1'
      },
      col2: {
        Type: TextCell,
        width: 100,
        label: 'Column 2'
      },
      col3: {
        Type: TextCell,
        width: 100,
        label: 'Column 3'
      },
    };

    const tree = renderer.create(
      <Grid
        data={data}
        columns={columns}
        indexColumn="col2" />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
