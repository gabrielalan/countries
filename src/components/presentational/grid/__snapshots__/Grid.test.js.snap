// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should render correct number of headers, rows and columns 1`] = `
<section
  className="grid"
>
  <div
    className="grid__row grid__row--header"
  >
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      <button
        className="grid__row__column__header-label"
        onClick={[Function]}
        type="button"
      >
        Column 1
      </button>
      <span
        className="grid__row__column__resizer"
        draggable={true}
        onDrag={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      />
    </div>
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      <button
        className="grid__row__column__header-label"
        onClick={[Function]}
        type="button"
      >
        Column 2
      </button>
      <span
        className="grid__row__column__resizer"
        draggable={true}
        onDrag={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      />
    </div>
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      <button
        className="grid__row__column__header-label"
        onClick={[Function]}
        type="button"
      >
        Column 3
      </button>
      <span
        className="grid__row__column__resizer"
        draggable={true}
        onDrag={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      />
    </div>
  </div>
  <div
    className="grid__row"
  >
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      Test 1
    </div>
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      1234
    </div>
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      1,2,3
    </div>
  </div>
  <div
    className="grid__row"
  >
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      Test 2
    </div>
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      3421
    </div>
    <div
      className="grid__row__column"
      style={
        Object {
          "width": 100,
        }
      }
    >
      5,5,5
    </div>
  </div>
</section>
`;
